### PR TITLE
Aruha 1501 choose partition hila

### PR DESCRIPTION
**Pull request by v-stepanov**
# Implement the functionality to read from a specific partitions of a subscription

> Zalando ticket : ARUHA-1501

## Description
Implementation of functionality to allow users to select partitions to read from a subscription.
API/documentation is here https://github.com/zalando/nakadi/pull/834

## Review
- [x] Tests
- [x] Documentation
- [x] CHANGELOG

## Deployment Notes
We should closely monitor the consumption of users

### CODE REVIEW

**Comment by v-stepanov** [comment]
Ok, I already noticed that I forgot one thing to implement: if "auto-balance" session got all partitions stolen and there are not enough partitions to provide to it - such session should be closed. I will implement that tomorrow.

File: src/acceptance-test/java/org/zalando/nakadi/webservice/hila/HilaRebalanceAT.java [hunk 1]
```diff
@@ -125,6 +128,132 @@ public void whenRebalanceThenPartitionsAreEquallyDistributedAndCommittedOffsetsA
...
+
+        // start a stream requesting to read from partition 7
+        final TestStreamingClient directClient2 = new TestStreamingClient(URL, subscription.getId(), "",
+                Optional.empty(),
+                Optional.of("{\"partitions\":[" +
+                        "{\"event_type\":\"" + eventType.getName() + "\",\"partition\":\"7\"}]}"));
+        directClient2.start();
+
+        // wait for rebalance to finish
+        Thread.sleep(1000);
```
Comment (lmontrieux) [inline src/acceptance-test/java/org/zalando/nakadi/webservice/hila/HilaRebalanceAT.java]: This is a bit brittle, but I don't really see a way to waitFor() the end of the rebalance (as opposed to waiting for some time, hoping that the rebalance will be finished)

File: src/acceptance-test/java/org/zalando/nakadi/webservice/hila/SubscriptionAT.java [hunk 2]
```diff
@@ -388,6 +390,47 @@ public void whenStatsOnNotInitializedSubscriptionThanCorrectResponse() throws IO
         }
     }
 
+    @Test
+    public void whenStreamDuplicatePartitionsThenUnprocessableEntity() throws IOException {
+        final String et = createEventType().getName();
+        final Subscription s = createSubscriptionForEventType(et);
+
+        final String body = "{\"partitions\":[" +
+                "{\"event_type\":\"et1\",\"partition\":\"0\"}," +
```
Comment (lmontrieux) [inline src/acceptance-test/java/org/zalando/nakadi/webservice/hila/SubscriptionAT.java]: Shouldn't you use `et` instead of `"et1"`? (same next line)

File: src/acceptance-test/java/org/zalando/nakadi/webservice/hila/SubscriptionAT.java [hunk 3]
```diff
@@ -388,6 +390,47 @@ public void whenStatsOnNotInitializedSubscriptionThanCorrectResponse() throws IO
...
+                .post("/subscriptions/{sid}/events", s.getId())
+                .then()
+                .statusCode(HttpStatus.SC_UNPROCESSABLE_ENTITY)
+                .body(JSON_HELPER.matchesObject(Problem.valueOf(
+                        MoreStatus.UNPROCESSABLE_ENTITY,
+                        "Duplicated partition specified")));
+    }
+
+    @Test
+    public void whenStreamWrongPartitionsThenUnprocessableEntity() throws IOException {
```
Comment (lmontrieux) [inline src/acceptance-test/java/org/zalando/nakadi/webservice/hila/SubscriptionAT.java]: Maybe it's overkill, but do you think it would be useful to have a test that tries to get partition `N + K`, from an event type with `N` partitions (where `K > 0`)?

File: src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java [hunk 4]
```diff
@@ -127,19 +148,49 @@ public OutputStream getOutputStream() {
...
-            @RequestParam(value = "batch_flush_timeout", required = false, defaultValue = "30") final int batchTimeout,
+            @Nullable @RequestParam(value = "batch_flush_timeout", required = false) final Integer batchTimeout,
             @Nullable @RequestParam(value = "stream_timeout", required = false) final Long streamTimeout,
-            @Nullable
-            @RequestParam(value = "stream_keep_alive_limit", required = false) final Integer streamKeepAliveLimit,
+            @Nullable @RequestParam(value = "stream_keep_alive_limit", required = false) final Integer
+                    streamKeepAliveLimit,
             final HttpServletRequest request, final HttpServletResponse response, final Client client)
-            throws IOException {
+            throws IOException, UnprocessableEntityException {
```
Comment (lmontrieux) [inline src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java]: I'm not sure where IOException and UnprocessableEntityException are thrown, and what is the difference with the previous method, which only throws UnprocessableEntityException?

File: src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java [hunk 5]
```diff
@@ -41,25 +43,31 @@
...
-            @Nullable final Long streamTimeoutSeconds, @Nullable final Integer batchKeepAliveIterations,
-            final int maxUncommittedMessages, final long commitTimeoutMillis, final Client consumingClient)
-            throws UnprocessableEntityException {
-        if (batchLimitEvents > 0) {
-            this.batchLimitEvents = batchLimitEvents;
-        } else {
-            throw new UnprocessableEntityException("batch_limit can't be lower than 1");
+    private final List<EventTypePartition> partitions;
+
+    private StreamParameters(final UserStreamParameters userParameters, final long commitTimeoutMillis,
```
Comment (lmontrieux) [inline src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java]: Why isn't `commitTimeoutMillis` part of `UserStreamParameters`?

File: src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java [hunk 6]
```diff
@@ -41,25 +43,31 @@
...
+
+        this.batchLimitEvents = userParameters.getBatchLimit().orElse(1);
+        if (batchLimitEvents <= 0) {
+            throw new WrongStreamParametersException("batch_limit can't be lower than 1");
         }
-        this.streamLimitEvents = Optional.ofNullable(streamLimitEvents).filter(v -> v != 0);
-        this.batchTimeoutMillis = batchTimeoutMillis;
+
+        this.streamLimitEvents = userParameters.getStreamLimit().filter(v -> v != 0);
+
```
Comment (lmontrieux) [inline src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java]: This is a very minor point, but I don't understand why there are empty lines here (except between lines 67-68, and 70-71)

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionRebalancer.java [hunk 7]
```diff
@@ -28,15 +77,15 @@
                 activeSessionIds.stream().mapToInt(activeSessionWeights::get).toArray());
 
         // Stage 1. Select partitions that are not assigned to any EXISTING session.
-        final Set<Partition> toRebalance = Stream.of(currentPartitions)
+        final List<Partition> toRebalance = Stream.of(currentPartitions)
```
Comment (lmontrieux) [inline src/main/java/org/zalando/nakadi/service/subscription/SubscriptionRebalancer.java]: Why did you change this?

File: src/main/java/org/zalando/nakadi/service/subscription/state/StartingState.java [hunk 8]
```diff
@@ -54,15 +58,28 @@ public void onEnter() {
...
      */
-    private void createSubscriptionLocked() {
+    private void initializeStream() {
         final boolean subscriptionJustInitialized = initializeSubscriptionLocked(getZk(),
                 getContext().getSubscription(), getContext().getTimelineService(), getContext().getCursorConverter());
         if (!subscriptionJustInitialized) {
+            // check if amount of streams <= the total amount of partitions
+            final Collection<Session> sessions = getZk().listSessions();
             final Partition[] partitions = getZk().getTopology().getPartitions();
             if (getZk().listSessions().size() >= partitions.length) {
```
Comment (lmontrieux) [inline src/main/java/org/zalando/nakadi/service/subscription/state/StartingState.java]: You can reuse `sessions` here instead of calling `getZk().listSessions()` a second time

Comment (lmontrieux) [reply src/main/java/org/zalando/nakadi/service/subscription/state/StartingState.java]: see line 66

File: src/acceptance-test/java/org/zalando/nakadi/webservice/hila/SubscriptionAT.java (hunk 2 above)
Comment (v-stepanov) [reply src/acceptance-test/java/org/zalando/nakadi/webservice/hila/SubscriptionAT.java]: I think I didn't get the question.
This test tests that Nakadi checks if user specified duplicated partition, so I need to provide the same event-type:partition pair twice as an input.

File: src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java (hunk 5 above)
Comment (v-stepanov) [reply src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java]: Because it is not specified by the user.

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionRebalancer.java (hunk 7 above)
Comment (v-stepanov) [reply src/main/java/org/zalando/nakadi/service/subscription/SubscriptionRebalancer.java]: Before this change the rabalance could provide different output for the same input which made it difficult to write tests. And I think it's good in general to have a predictable result. I asked @antban - there was no intention to do randomization here so he blessed me to change it to List.

File: src/main/java/org/zalando/nakadi/service/subscription/state/StartingState.java (hunk 8 above)
Comment (v-stepanov) [reply src/main/java/org/zalando/nakadi/service/subscription/state/StartingState.java]: good catch; fixed https://github.com/zalando/nakadi/pull/848/commits/74026cef64c5118980542664e8d6c5b7bf25e316

File: src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java (hunk 6 above)
Comment (v-stepanov) [reply src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java]: I just wanted to improve readability using empty lines :) but it seems that it brings confusion - so I will remove empty lines.

**Comment by v-stepanov** [comment]
I added now functionality to close streams which loose all partitions

File: src/acceptance-test/java/org/zalando/nakadi/webservice/hila/SubscriptionAT.java (hunk 3 above)
Comment (v-stepanov) [reply src/acceptance-test/java/org/zalando/nakadi/webservice/hila/SubscriptionAT.java]: Yes, makes sense. I added that https://github.com/zalando/nakadi/pull/848/commits/b982da9499e7c05ee6672317178942fc5f72ee2c

File: src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java (hunk 4 above)
Comment (v-stepanov) [reply src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java]: Yes, these exceptions are not thrown actually, it remained from some old times. I removed that now https://github.com/zalando/nakadi/pull/848/commits/220a7106992d5125c9f027b29001b02ea750f7d4

File: src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java (hunk 6 above)
Comment (v-stepanov) [reply src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java]: https://github.com/zalando/nakadi/pull/848/commits/cafbbecadb02e7908b7eeecd47979c9bcd85d1f6

File: src/acceptance-test/java/org/zalando/nakadi/webservice/hila/HilaRebalanceAT.java (hunk 1 above)
Comment (v-stepanov) [reply src/acceptance-test/java/org/zalando/nakadi/webservice/hila/HilaRebalanceAT.java]: I think it's possible to check that using stats endpoint. I will check if I can use that.

Comment (v-stepanov) [reply src/acceptance-test/java/org/zalando/nakadi/webservice/hila/HilaRebalanceAT.java]: Done https://github.com/zalando/nakadi/pull/848/commits/ec3f4429aabf2181e0f1ffbb5f5643dce9fe9970

**Comment by v-stepanov** [comment]
@lmontrieux I addressed all your comments

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionRebalancer.java (hunk 7 above)
Comment (lmontrieux) [reply src/main/java/org/zalando/nakadi/service/subscription/SubscriptionRebalancer.java]: Ok, I agree that it is good to have predictable test results.

File: src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java (hunk 5 above)
Comment (lmontrieux) [reply src/main/java/org/zalando/nakadi/service/subscription/StreamParameters.java]: Oh yes, of course.

**Comment by lmontrieux** [comment]
👍

**Comment by v-stepanov** [comment]
👍

**Comment by v-stepanov** [comment]
deploy validation please

File: src/main/java/org/zalando/nakadi/service/subscription/IdleStreamWatcher.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
+package org.zalando.nakadi.service.subscription;
+
+public class IdleStreamWatcher {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/subscription/IdleStreamWatcher.java]: it is not clear why this class is synchronised at all, it is used in StreamingState, which runs in one thread.

File: src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java [hunk 10]
```diff
@@ -257,7 +258,13 @@ public final void unregisterSession(final Session session) {
         return loadDataAsync(
                 zkSessions,
                 key -> getSubscriptionPath("/sessions/" + key),
-                (key, data) -> new Session(key, Integer.parseInt(new String(data, UTF_8)))).values();
+                (key, data) -> {
+                    try {
+                        return deserializeSession(key, data);
+                    } catch (final IOException e) {
+                        throw new NakadiRuntimeException(e);
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java]: why not to put it on the level of deserialisation?

File: src/main/java/org/zalando/nakadi/service/subscription/IdleStreamWatcher.java (hunk 9 above)
Comment (v-stepanov) [reply src/main/java/org/zalando/nakadi/service/subscription/IdleStreamWatcher.java]: I think it doesn't always run in one Thread. It has code which is executed from ZK-Watcher threads.

File: src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java (hunk 10 above)
Comment (v-stepanov) [reply src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java]: I think that deserialization that throws IOException is more logical. So that the callers of that should handle the possible exception.

Comment (adyach) [reply src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java]: we just throw sand in the wheels by ourselves, I am more than sure will process it the same place in other places, because you can not do much with it having any exception type.

**Comment by lmontrieux** [comment]
This addition looks fine to me, so if it is ok with @adyach I think it can be approved, after you solve the confilct.

**Comment by adyach** [comment]
LGTM

**Comment by lmontrieux** [comment]
👍

**Comment by lmontrieux** [comment]
Deploy validation please

**Comment by v-stepanov** [comment]
deploy validation please

**Comment by lmontrieux** [comment]
👍

**Comment by v-stepanov** [comment]
deploy validation please

**Comment by lmontrieux** [comment]
👍

**Comment by v-stepanov** [comment]
👍

**Comment by lmontrieux** [comment]
👍
